File: Groupe1/TP2/src/chaine.c
#include <stdio.h>

// Fonction pour calculer la longueur d'une chaîne de caractères
int longueur_chaine(const char* str) {
    int longueur = 0;
    // Boucle tant que le caractère courant n'est pas '\0' (le caractère nul)
    while (str[longueur] != '\0') {
        longueur++;  // Incrémenter la longueur pour chaque caractère
    }
    return longueur;  // Retourner la longueur de la chaîne
}

// Fonction pour copier une chaîne dans une autre
void copie_chaine(char* destination, const char* source) {
    int i = 0;
    // Boucle pour copier chaque caractère de la source vers la destination
    while (source[i] != '\0') {
        destination[i] = source[i];  // Copier le caractère de source à destination
        i++;
    }
    destination[i] = '\0';  // Ajouter le caractère nul à la fin de la chaîne copiée
}

// Fonction pour concaténer deux chaînes
void concatene_chaine(char* destination, const char* source1, const char* source2) {
    int i = 0, j = 0;
    
    // Copier la première chaîne (source1) dans destination
    while (source1[i] != '\0') {
        destination[i] = source1[i];  // Ajouter chaque caractère de source1 à destination
        i++;
    }
    
    // Ajouter la deuxième chaîne (source2) à la suite de destination
    while (source2[j] != '\0') {
        destination[i + j] = source2[j];  // Ajouter chaque caractère de source2 à destination
        j++;
    }
    
    destination[i + j] = '\0';  // Ajouter le caractère nul à la fin de la chaîne concaténée
}

int main() {
    // Déclaration des chaînes de caractères avec des tailles maximales
    char chaine1[100], chaine2[100], resultat[200];  // `resultat` doit être assez grand pour contenir les deux chaînes concaténées

    // Demander à l'utilisateur de saisir deux chaînes
    printf("Entrez la première chaîne : ");
    // Utilisation de `scanf` pour lire la première chaîne, limité à 99 caractères pour éviter le dépassement de buffer
    scanf("%99s", chaine1);
    printf("Entrez la deuxième chaîne : ");
    // Utilisation de `scanf` pour lire la deuxième chaîne, limité à 99 caractères également
    scanf("%99s", chaine2);

    // Calcul et affichage de la longueur des deux chaînes
    printf("Longueur de la première chaîne : %d\n", longueur_chaine(chaine1));  // Affiche la longueur de `chaine1`
    printf("Longueur de la deuxième chaîne : %d\n", longueur_chaine(chaine2));  // Affiche la longueur de `chaine2`

    // Copie de la première chaîne dans `resultat`
    copie_chaine(resultat, chaine1); //appel de la fonction "copie_chaine" pour copier la chaine1 donc le resultat sera stocké dans la variable "copie"
    printf("Copie de la première chaîne : %s\n", resultat);  // Affichage du résultat de la copie

    // Concaténation des deux chaînes dans `resultat`
    concatene_chaine(resultat, chaine1, chaine2); // appel de la fonction "concatene_chaine" pour Concaténer les deux chaînes donc le resultat sera stocké dans la variable "concatenation"
    printf("Concaténation des deux chaînes : %s\n", resultat);   //Affichage du résultat de la concaténation

    return 0;
}

File: Groupe1/TP2/src/chaine.py
# Fonction pour calculer la longueur d'une chaîne de caractères
def longueur_chaine(chaine):
    longueur = 0  # Initialisation de la variable longueur à 0
    # Parcourir chaque caractère dans la chaîne
    for caractere in chaine:
        longueur += 1  # Incrémenter la longueur pour chaque caractère
    return longueur  # Retourner la longueur totale

# Fonction pour copier une chaîne de caractères dans une autre
def copie_chaine(source):
    destination = ''  # Initialisation de la chaîne vide qui recevra la copie
    # Parcourir chaque caractère de la chaîne source
    for caractere in source:
        destination += caractere  # Ajouter chaque caractère à la chaîne destination
    return destination  # Retourner la chaîne copiée

# Fonction pour concaténer deux chaînes de caractères
def concatene_chaine(source1, source2):
    resultat = ''  # Initialisation de la chaîne résultat vide
    
    # Copier la première chaîne dans la chaîne résultat
    for caractere in source1:
        resultat += caractere  # Ajouter chaque caractère de source1 dans resultat
    
    # Ajouter la deuxième chaîne à la fin de la première
    for caractere in source2:
        resultat += caractere  # Ajouter chaque caractère de source2 dans resultat
    
    return resultat  # Retourner la chaîne concaténée

if __name__ == "__main__":  # Le point d'entrée du programme
    # Demander à l'utilisateur de saisir deux chaînes
    chaine1 = input("Entrez la première chaîne : ")  # Saisie de la première chaîne
    chaine2 = input("Entrez la deuxième chaîne : ")  # Saisie de la deuxième chaîne

    # Calcul et affichage de la longueur des chaînes
    print("Longueur de la première chaîne :", longueur_chaine(chaine1))  # Afficher la longueur de chaine1
    print("Longueur de la deuxième chaîne :", longueur_chaine(chaine2))  # Afficher la longueur de chaine2

    # Copie de la première chaîne et affichage
    copie = copie_chaine(chaine1)  # appel de la fonction "copie_chaine" pour copier la chaine1 donc le resultat sera stocké dans la variable "copie"
    print("Copie de la première chaîne :", copie)  # Affichage du résultat de la copie

    # Concaténation des deux chaînes et affichage
    concatenation = concatene_chaine(chaine1, chaine2)  # appel de la fonction "concatene_chaine" pour Concaténer les deux chaînes donc le resultat sera stocké dans la variable "concatenation"
    print("Concaténation des deux chaînes :", concatenation)  # Affichage du résultat de la concaténation
